### PR TITLE
ci: Update Clang in "tidy" job

### PR DESCRIPTION
**Pull request by hebasto**
This PR switches to the latest [IWYU 0.23](https://github.com/include-what-you-use/include-what-you-use/releases/tag/0.23), which is compatible with Clang 19.

New "bugprone-use-after-move" and "modernize-use-starts-ends-with" warnings that emerged have been addressed.

### CODE REVIEW

File: src/.clang-tidy [hunk 1]
```diff
@@ -19,8 +19,10 @@ performance-*,
 -performance-avoid-endl,
 -performance-enum-size,
 -performance-inefficient-string-concatenation,
+-performance-inefficient-vector-operation,
```
Comment (l0rinc) [inline src/.clang-tidy:22]: Given that https://github.com/bitcoin/bitcoin/pull/31305 fixes the issue, are we disabling it as a temporary measure?

Comment (hebasto) [reply src/.clang-tidy:22]: Right.

It depends on which PR is merged first :)

File: src/test/coins_tests.cpp [hunk 2]
```diff
@@ -771,11 +771,10 @@ static void CheckAddCoinBase(CAmount base_value, CAmount cache_value, CAmount mo
 // while still verifying that the CoinsViewCache::AddCoin implementation
 // ignores base values.
 template <typename... Args>
-static void CheckAddCoin(Args&&... args)
```
Comment (l0rinc) [inline src/test/coins_tests.cpp:774]: This was already removed in the ACK-ed https://github.com/bitcoin/bitcoin/pull/30906/files#diff-3d0856e8b7f136c588b229e0cbd3b2e2c309cd097ade0879521daba4e1bb2a33L774-L777, could you please either revert this line or do it in a separate PR, since I think removing it completely is a better fix anyway.

File: src/test/result_tests.cpp [hunk 3]
```diff
@@ -59,12 +59,11 @@ void ExpectResult(const util::Result<T>& result, bool success, const bilingual_s
 }
 
 template <typename T, typename... Args>
-void ExpectSuccess(const util::Result<T>& result, const bilingual_str& str, Args&&... args)
+void ExpectSuccess(const util::Result<T>& result, const bilingual_str& str, const Args&... args)
```
Comment (l0rinc) [inline src/test/result_tests.cpp:62]: Looking at the actual usage, `ExpectSuccess` doesn't need to be a variadic template function:
```C++
ExpectSuccess(IntFn(5, true), {}, 5);
ExpectSuccess(NoCopyFn(5, true), {}, 5);
ExpectSuccess(StrFn(Untranslated("S"), true), {}, Untranslated("S"));
```
, we might as well simplify to something like:
```C++
template <typename T, typename U>
void ExpectSuccess(const util::Result<T>& result, const bilingual_str& str, U&& expected)
{
    ExpectResult(result, true, str);
    BOOST_CHECK_EQUAL(result.has_value(), true);
    BOOST_CHECK_EQUAL(result.value(), expected);
    BOOST_CHECK_EQUAL(&result.value(), &*result);
}
```

**Review by l0rinc — CHANGES_REQUESTED** [review]
Concept ACK, but approach NACK, I think some of the examples can be fixed by removing the "feature" instead of just patching them.

File: ci/test/00_setup_env_native_tidy.sh [hunk 4]
```diff
@@ -6,9 +6,9 @@
 
 export LC_ALL=C.UTF-8
 
-export CI_IMAGE_NAME_TAG="docker.io/ubuntu:24.04"
+export CI_IMAGE_NAME_TAG="docker.io/ubuntu:24.10"
```
[the remark this answers is not in the document]
Comment (hebasto) [reply ci/test/00_setup_env_native_tidy.sh]: Thanks! Switched to use `APT_LLVM_V`.

File: src/test/result_tests.cpp [hunk 5]
```diff
@@ -59,11 +59,11 @@ void ExpectResult(const util::Result<T>& result, bool success, const bilingual_s
...
 {
     ExpectResult(result, true, str);
     BOOST_CHECK_EQUAL(result.has_value(), true);
-    BOOST_CHECK_EQUAL(result.value(), T{std::forward<Args>(args)...});
+    BOOST_CHECK_EQUAL(result.value(), T{args...});
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp]: This change has been removed from this PR.

File: src/test/coins_tests.cpp (hunk 2 above)
Comment (hebasto) [reply src/test/coins_tests.cpp:774]: This change has been removed from this PR.

File: src/test/result_tests.cpp (hunk 3 above)
Comment (hebasto) [reply src/test/result_tests.cpp]: This change has been removed from this PR.

File: src/test/coins_tests.cpp (hunk 2 above)
Comment (l0rinc) [reply src/test/coins_tests.cpp:774]: Thank you!

**Review by l0rinc — APPROVED** [review]
ACK 3e039b7a03bea6e11e994aebe038ebd6ca80aeb8

The latest change:
* Reverted the `CI_IMAGE_NAME_TAG` and added `APT_LLVM_V`.
* Reverted the `CheckAddCoin` and `ExpectSuccess` fixes to avoid collisions with other PRs.
* Updated description to reflect the dependent fix PRs.

File: src/wallet/walletdb.cpp [hunk 6]
```diff
@@ -1013,7 +1013,7 @@ static DBErrors LoadAddressBookRecords(CWallet* pwallet, DatabaseBatch& batch) E
             // "1" or "p" for present (which was written prior to
             // f5ba424cd44619d9b9be88b8593d69a7ba96db26).
             pwallet->LoadAddressPreviouslySpent(dest);
-        } else if (strKey.compare(0, 2, "rr") == 0) {
+        } else if (strKey.compare(0, 2, "rr") == 0) {  // NOLINT(modernize-use-starts-ends-with)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/wallet/walletdb.cpp]: > Seems easier to just fix it right away, instead of touching the line to say to fix it and then fix it in the next commit?

The commits have been squashed.

> But I'd like to understand the reason for the supressions: would the lint problem break CI without them?

Yes, it would.

File: src/test/result_tests.cpp [hunk 7]
```diff
@@ -63,6 +63,7 @@ void ExpectSuccess(const util::Result<T>& result, const bilingual_str& str, Args
 {
     ExpectResult(result, true, str);
     BOOST_CHECK_EQUAL(result.has_value(), true);
+    // NOLINTNEXTLINE(bugprone-use-after-move)
```
[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp:66]: Right. I outlined this issue for follow-ups in the PR description.

[the remark this answers is not in the document]
Comment (l0rinc) [reply src/test/result_tests.cpp:66]: We can likely simplify the call here, see: https://github.com/bitcoin/bitcoin/pull/31306#discussion_r1846343700

[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp:66]: Fixing all new "bugprone-use-after-move" warnings contradicts with other [comments](https://github.com/bitcoin/bitcoin/pull/31306#discussion_r1846324884). On the other hand, fixing some of them is not consistent. That's why I chose to silence them.

`NOLINT` comments are easy searchable and effectively document the issues that have to be resolved.

[the remark this answers is not in the document]
Comment (l0rinc) [reply src/test/result_tests.cpp:66]: Unfortunately silencing will also invalidate the ACK-ed PR.
But if we could instead add `-bugprone-use-after-move` to `src/.clang-tidy` this concern could be addressed separately.

[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp:66]: It will expose the whole codebase to re-introducing related bugs, no?

[the remark this answers is not in the document]
Comment (l0rinc) [reply src/test/result_tests.cpp:66]: We could tackle it in a separate pr like we do with the `performance-*` checks, right?

[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp:66]: We could, but `performance-*` checks are not about safety, whereas `bugprone-*` checks are.

[the remark this answers is not in the document]
Comment (l0rinc) [reply src/test/result_tests.cpp:66]: If you decide to fix them here, I'll just rebase the related PR and ask for a re-review, it's not a tragedy :)

[the remark this answers is not in the document]
Comment (theuni) [reply src/test/result_tests.cpp:66]: > NOLINT comments are easy searchable and effectively document the issues that have to be resolved.

Sorry, but no, that is exactly the _opposite_ if what `NOLINT` documents!

`NOLINT` means "I'm smarter than my static analyzer".

`TODO` would effectively document an issue to be resolved.

Hard NACK here due to the reasoning being applied. In fact, it rises to the level of suspicion. I'll be reviewing all of these refactors in great detail.

[the remark this answers is not in the document]
Comment (l0rinc) [reply src/test/result_tests.cpp:66]: @hebasto, what if we invert this and push the fixes in separate PRs and only merge this TIDY_LLVM_V upgrade after all problems were already resolved?

[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp:66]: > > NOLINT comments are easy searchable and effectively document the issues that have to be resolved.
> 
> Sorry, but no, that is exactly the _opposite_ of what `NOLINT` documents!
> 
> `NOLINT` means "I'm smarter than my static analyzer".
> 
> `TODO` would effectively document an issue to be resolved.
> 
> Hard NACK here due to the reasoning being applied. In fact, it rises to the level of suspicion. I'll be reviewing all of these refactors in great detail.

A bare `NOLINT`, without any explanation, has never meant "I'm smarter than my static analyzer"—at least, not to me.

File: src/test/coins_tests.cpp [hunk 8]
```diff
@@ -773,8 +773,10 @@ static void CheckAddCoinBase(CAmount base_value, CAmount cache_value, CAmount mo
 template <typename... Args>
 static void CheckAddCoin(Args&&... args)
 {
-    for (const CAmount base_value : {ABSENT, SPENT, VALUE1})
+    for (const CAmount base_value : {ABSENT, SPENT, VALUE1}) {
+        // NOLINTNEXTLINE(bugprone-use-after-move)
```
Comment (theuni) [inline src/test/coins_tests.cpp:777]: As mentioned [here](https://github.com/bitcoin/bitcoin/pull/31364#discussion_r1857161138), this change struck me as dangerous.

> It shouldn't be marked nolint, rather it should be fixed: drop the forward and pass by const Args&.

I understand the desire to turn on new checks and notate the false-positives, but the result should _always_ be better/safer/cleaner code.

In this case, the result would've been a latent bug explicitly marked notabug. And while a buggy test isn't the end of the world, the suggestion that the approach for recent similar PR's might have been "quiet the static analysis tool first, fix the bug later" struck me as so backwards that I had to wonder if any other actual bugs in more important code had been masked as well. Looking through recent/current PRs, I don't see any other examples of that (modulo the boost issue pointed out here).

Please prioritize correctness and consider what you're signaling to reviewers.

Comment (hebasto) [reply src/test/coins_tests.cpp:777]: Right.

Comment (hebasto) [reply src/test/coins_tests.cpp:777]: > As mentioned [here](https://github.com/bitcoin/bitcoin/pull/31364#discussion_r1857161138), this change struck me as dangerous.

My apologies.

An additional context available here: https://github.com/bitcoin/bitcoin/pull/31306#discussion_r1846324884.

File: src/test/result_tests.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (theuni) [reply src/test/result_tests.cpp:66]: > A bare NOLINT, without any explanation, has never meant "I'm smarter than my static analyzer"—at least, not to me.

I'm not sure what else it could mean.

[From the docs](https://clang.llvm.org/extra/clang-tidy/): "clang-tidy diagnostics are intended to call out code that does not adhere to a coding standard, or is otherwise problematic in some way. However, if the code is known to be correct, it may be useful to silence the warning"

Does this mean there are other places in the code where `NOLINT` is used as a means of documenting a todo? If so, we need to address those asap.

[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp:66]: > I'm not sure what else it could mean.
> 
> [From the docs](https://clang.llvm.org/extra/clang-tidy/): "clang-tidy diagnostics are intended to call out code that does not adhere to a coding standard, or is otherwise problematic in some way. However, if the code is known to be correct, it may be useful to silence the warning"

You are right.

[the remark this answers is not in the document]
Comment (theuni) [reply src/test/result_tests.cpp:66]: @hebasto No need to close this if we can change the approach. I think the changes are useful (once the actual bugs are fixed), but your suggestion that other real bugs might have been previously masked set off my alarm bells. Looking through the code now (bench/tests excluded), I don't see any other examples of this.

I think we can agree on an approach where we (in this order):
- Fix the actual bugs
- Notate the false-positives
- Enable the checks.

It sounds like it also is worth adding a line in `developer-notes.md` that describes when a `NOLINT` is/is not appropriate.

**Comment by hebasto** [comment]
> No need to close this if we can change the approach. 

Rebased on top of the merged #31305 and #31364.

> I think we can agree on an approach where we (in this order):
> 
>    * Fix the actual bugs
> 
>    * Notate the false-positives
> 
>    * Enable the checks.

Done.

> It sounds like it also is worth adding a line in `developer-notes.md` that describes when a `NOLINT` is/is not appropriate.

Left this change for a follow-up.

File: src/test/coins_tests.cpp (hunk 2 above)
Comment (l0rinc) [reply src/test/coins_tests.cpp:774]: I'd like to revive my original concerns here: [`c4eb31b` (#30906)](https://github.com/bitcoin/bitcoin/pull/30906/commits/c4eb31be3e56936479921b7556fcba1747867d02#diff-3d0856e8b7f136c588b229e0cbd3b2e2c309cd097ade0879521daba4e1bb2a33L778-L782) removes this method completely, which I think is a better fix for it than the current workaround.

File: src/test/result_tests.cpp [hunk 9]
```diff
@@ -63,6 +63,12 @@ void ExpectSuccess(const util::Result<T>& result, const bilingual_str& str, Args
 {
     ExpectResult(result, true, str);
     BOOST_CHECK_EQUAL(result.has_value(), true);
+    // Boost.Test versions < 1.87 contain a bug that triggers
```
Comment (l0rinc) [inline src/test/result_tests.cpp:66]: I'm not sure I understand the reason for the suppression, please see my original suggestion in https://github.com/bitcoin/bitcoin/pull/31306#discussion_r1846343700

**Comment by hebasto** [comment]
Rebased on top of the merged https://github.com/bitcoin/bitcoin/pull/30906.

File: src/test/coins_tests.cpp (hunk 2 above)
Comment (hebasto) [reply src/test/coins_tests.cpp]: https://github.com/bitcoin/bitcoin/pull/30906 has been merged, which makes this discussion outdated.

**Review by maflcko — APPROVED** [review]
lgtm, but it would be better to apply my suggestion. Otherwise the test code will have to be touched again at some point in the future.

File: src/test/result_tests.cpp (hunk 7 above)
[the remark this answers is not in the document]
Comment (hebasto) [reply src/test/result_tests.cpp]: > Have you seen this suggestion?

Thanks! Your suggestion has been accepted.

**Comment by hebasto** [comment]
> lgtm, but it would be better to apply my suggestion. Otherwise the test code will have to be touched again at some point in the future.

The suggestion has been applied.

**Comment by l0rinc** [comment]
ACK 31e59d94c67b58f24dd701fc7ccfee7d79f311cb

File: src/test/coins_tests.cpp (hunk 2 above)
Comment (l0rinc) [reply src/test/coins_tests.cpp]: Yes, thanks for waiting it out

**Review by theuni — APPROVED** [review]
ACK 31e59d94c67b58f24dd701fc7ccfee7d79f311cb
